File: Hash-Based Searching/Locality Sensitive Hashing (LSH)/LocalitySensitiveHashingLSH.c
#include <stdio.h>
#include <stdlib.h>

#define DIMENSIONS 4
#define HYPERPLANES 3

typedef struct {
    double values[DIMENSIONS];
} Vector;

double dotProduct(
    Vector a,
    Vector b
) {
    double result = 0.0;

    for (
        int i = 0;
        i < DIMENSIONS;
        i++
    ) {
        result +=
            a.values[i] *
            b.values[i];
    }

    return result;
}

int computeHash(
    Vector point,
    Vector planes[]
) {

    int hash = 0;

    for (
        int i = 0;
        i < HYPERPLANES;
        i++
    ) {

        if (
            dotProduct(
                point,
                planes[i]
            ) >= 0
        ) {
            hash |= (1 << i);
        }
    }

    return hash;
}

int main() {

    Vector hyperplanes[
        HYPERPLANES
    ] = {
        {{1, 0, -1, 0}},
        {{0, 1, 0, -1}},
        {{1, 1, 1, 1}}
    };

    Vector point1 =
        {{2, 3, 1, 0}};

    Vector point2 =
        {{2, 2, 1, 1}};

    int hash1 =
        computeHash(
            point1,
            hyperplanes
        );

    int hash2 =
        computeHash(
            point2,
            hyperplanes
        );

    printf(
        "Point1 Hash: %d\n",
        hash1
    );

    printf(
        "Point2 Hash: %d\n",
        hash2
    );

    if (hash1 == hash2)
        printf(
            "Likely Similar\n"
        );
    else
        printf(
            "Likely Different\n"
        );

    return 0;
}

File: Hash-Based Searching/Locality Sensitive Hashing (LSH)/LocalitySensitiveHashingLSH.py
"""
LocalitySensitiveHashingLSH.py
Cosine Similarity LSH using Random Hyperplanes
"""

DIMENSIONS = 4
HYPERPLANES = 3


class Vector:

    def __init__(self, values):
        self.values = values


def dot_product(a, b):

    result = 0.0

    for i in range(DIMENSIONS):

        result += (
            a.values[i]
            * b.values[i]
        )

    return result


def compute_hash(
    point,
    hyperplanes
):

    hash_value = 0

    for i in range(HYPERPLANES):

        if (
            dot_product(
                point,
                hyperplanes[i]
            )
            >= 0
        ):

            hash_value |= (
                1 << i
            )

    return hash_value


if __name__ == "__main__":

    hyperplanes = [

        Vector(
            [1, 0, -1, 0]
        ),

        Vector(
            [0, 1, 0, -1]
        ),

        Vector(
            [1, 1, 1, 1]
        )
    ]

    point1 = Vector(
        [2, 3, 1, 0]
    )

    point2 = Vector(
        [2, 2, 1, 1]
    )

    hash1 = compute_hash(
        point1,
        hyperplanes
    )

    hash2 = compute_hash(
        point2,
        hyperplanes
    )

    print(
        "Point1 Hash:",
        hash1
    )

    print(
        "Point2 Hash:",
        hash2
    )

    if hash1 == hash2:
        print(
            "Likely Similar"
        )
    else:
        print(
            "Likely Different"
        )
